### PR TITLE
[dask] Add support for 'pred_leaf' in Dask estimators (fixes #3792)

### LINKED ISSUE #3792: [dask] support 'pred_leaf' in predict()
## Summary

See https://github.com/microsoft/LightGBM/pull/3774#issuecomment-763136825 for background.

To close this issue, add tests to https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_dask.py confirming that `.predict(X, pred_leaf=True)` works for `DaskLGBMClassifier` and `DaskLGBMRegressor` .

## Motivation

Adding this feature would allow users to understand which leaves in the trained model's trees their records fall into. This would bring the Dask interface closer to parity with `lightgbm.sklearn`.

## References

When I first attempted this in #3774, I hit some issues related to how LightGBM uses `dask.DataFrame.map_partitions()` and `dask.Array.map_blocks()`. These function allow you to apply a function to each partition of a distributed data collection, like `X.map(some_func)`. In some situations, they require you to provide additional information about the shape and data types of the return from `some_func()`.

You can learn more about this in the `map_partitions()` docs: https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.map_partitions

The results from `predict(X, pred_leaf=True)` should be an array with shape (`num_rows_in_X`, `num_trees_in_model`). Note that for multi-class classification, `num_trees` will be `n_estimators * num_classes`.

For testing this feature, you may find https://github.com/jameslamb/lightgbm-dask-testing useful.

### PR DESCRIPTION
**Pull request by jameslamb**
This PR adds support in the Dask module for getting indices of the leaf nodes in each tree that each observation in a dataset falls into, via `.predict(X, pred_leaf=True)`. It fixes #3792.

This brings the Dask module a step closer to feature parity with the estimators available from `lightgbm.sklearn`.

## Notes for Reviewers

These tests also led to the discovery of this bug: https://github.com/microsoft/LightGBM/issues/3918

### CODE REVIEW

File: tests/python_package_test/test_dask.py [hunk 1]
```diff
@@ -241,6 +242,17 @@ def test_classifier(output, centers, client, listen_port):
     assert_eq(p1_local, p2)
     assert_eq(y, p1_local)
 
+    # pref_leaf values should have the right shape
+    # and values that look like valid tree nodes
+    assert p1_pred_leaf.compute().shape == (
+        X.shape[0],
+        dask_classifier.booster_.num_trees()
+    )
+    pred_leaf_vals = p1_pred_leaf.compute()
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:251]: Can `.compute()` be called only once?

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:251]: oh yep, good point!

File: tests/python_package_test/test_dask.py [hunk 2]
```diff
@@ -241,6 +242,17 @@ def test_classifier(output, centers, client, listen_port):
     assert_eq(p1_local, p2)
     assert_eq(y, p1_local)
 
+    # pref_leaf values should have the right shape
+    # and values that look like valid tree nodes
+    assert p1_pred_leaf.compute().shape == (
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:247]:
```suggestion
    pred_leaf_vals = p1_pred_leaf.compute()
    assert pred_leaf_vals.shape == (
```

File: tests/python_package_test/test_dask.py [hunk 3]
```diff
@@ -399,6 +413,17 @@ def test_regressor(output, client, listen_port):
     # Predictions should be roughly the same.
     assert_eq(p1, p1_local)
 
+    # pref_leaf values should have the right shape
+    # and values that look like valid tree nodes
+    assert p1_pred_leaf.compute().shape == (
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:418]:
```suggestion
    pred_leaf_vals = p1_pred_leaf.compute()
    assert pred_leaf_vals.shape == (
```

File: tests/python_package_test/test_dask.py [hunk 4]
```diff
@@ -595,6 +621,17 @@ def test_ranker(output, client, listen_port, group):
     assert spearmanr(rnkvec_dask, rnkvec_local).correlation > 0.8
     assert_eq(rnkvec_dask, rnkvec_dask_local)
 
+    # pref_leaf values should have the right shape
+    # and values that look like valid tree nodes
+    assert p1_pred_leaf.compute().shape == (
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:626]:
```suggestion
    pred_leaf_vals = p1_pred_leaf.compute()
    assert pred_leaf_vals.shape == (
```

File: tests/python_package_test/test_dask.py [hunk 5]
```diff
@@ -595,6 +621,17 @@ def test_ranker(output, client, listen_port, group):
...
+        X.shape[0],
+        dask_ranker.booster_.num_trees()
+    )
+    pred_leaf_vals = p1_pred_leaf.compute()
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:630]:
```suggestion
```

File: tests/python_package_test/test_dask.py [hunk 6]
```diff
@@ -399,6 +413,17 @@ def test_regressor(output, client, listen_port):
     # Predictions should be roughly the same.
     assert_eq(p1, p1_local)
 
+    # pref_leaf values should have the right shape
+    # and values that look like valid tree nodes
+    assert p1_pred_leaf.compute().shape == (
+        X.shape[0],
+        dask_regressor.booster_.num_trees()
+    )
+    pred_leaf_vals = p1_pred_leaf.compute()
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:422]:
```suggestion
```

File: tests/python_package_test/test_dask.py (hunk 1 above)
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:251]:
```suggestion
```

**Review by StrikerRUS — APPROVED** [review]
Nice! Thank you!

**Pull request merged**